engines/bops_submissions/app/services/bops_submissions/zip_extraction_service.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+require "zip"
+require "uri"
+require "faraday"
+
+module BopsSubmissions
+  class ZipExtractionService
+    OPEN_TIMEOUT = 10
+    READ_TIMEOUT = 20
+
+    ATTACHABLE_EXTENSIONS = %w[.pdf .docx .doc .jpg .jpeg .png].freeze
+    APPLICATION_FILE = "Application.json"
+
+    def initialize(submission:)
+      @submission = submission
+    end
+
+    attr_reader :submission
+
+    def call
+      submission.document_link_urls.each do |url|
+        process_with_zip_input_stream(url)
+      rescue Zip::Error => e
```

Comment:
Do we know why Zip::InputStream sometimes fails when Zip::File works? Is it simpler always to rely on Zip::File or are there other downsides to that?

Reply:
Oh never mind, I see it's in the second commit message 👍🏻

Reply:
It's a little annoying that these two `extract_*` methods are kind of similar and yet not similar enough.

It'd be ideal if we were able to do something like `def extract(url, extractor_class)` and call it like `extract(url, Zip::InputStream)` or `extract(url, Zip::File)`, but it looks like the behaviour of the two classes is different in a few places. Oh well, can't be helped.

Reply:
Yeah I was experimenting abstracting and using a yield type approach but think it's just different enough that it makes sense to be separated

Reply:
Streaming will be more efficient but there may be some zip archives where you have to load it all to decompress